TELCODOCS-640: Optimizing CPU usage (#51562)
openshift/openshift-docs: modules/supporting_encapsulation.adoc

// Module included in the following assemblies:
//
// * scalability_and_performance/optimizing-cpu-usage.adoc

[id="supporting-encapsulation_{context}"]

> = Supporting encapsulation

Reviewer: The purpose of this section is unclear
Suggest this is clarified or section skipped

Reviewer 2: FYI @djuran Tony is unexpectedly out sick, I will take over this PR and complete it in https://github.com/openshift/openshift-docs/pull/53033

// Module included in the following assemblies:
//
// * scalability_and_performance/optimizing-cpu-usage.adoc

[id="supporting-encapsulation_{context}"]
= Supporting encapsulation

As an OpenShift support engineer or developer, you want to inspect Kubernetes-specific mountpoints.

== Locating the encapsulated namespaces

> Any monitoring tool that relies on the ability to run in the host OS and have visibility of mountpoints created by Kubelet, the Container Runtime, or containers themselves, will need to enter the container mount namespace in order to see these mountpoints.

Reviewer 3: ```suggestion
Any monitoring tool that relies on the ability to run in the host OS and have visibility of mountpoints created by Kubelet, the Container Runtime, or containers themselves, will need to enter the container mount namespace in order to see these mountpoints. The provided `kubensenter` script will execute another command inside the Kubernetes mountpoint and can be used to adapt any existing tools.

For example, if a systemd service needs to run inside the new Kubernetes mount namespace, you should edit the service file and prefix the `ExecStart=` command line with `kubensenter`.
```

Author: Done.

> [source,yaml]

Reviewer 3: ```suggestion
[source,terminal]
```

Author: Not afaik. Using terminal.

----

>   sh-4.4# readlink /proc/$(pgrep kubelet)/ns/mnt
> 
>   sh-4.4# readlink /proc/$(pgrep crio)/ns/mnt

Reviewer 3: ```suggestion
[Unit]
Description=Sample service

[Service]
ExecStart=/usr/bin/kubensenter /path/to/original/command arg1 arg2
```

Author: Done.

----

== Inspecting the encapsulated namespaces

> An administrator logged in to the host OS wishing to obtain information about the Kubernetes system for debugging or auditing purposes will need to be aware of the new mount namespace, and will need to run `nsenter` to view or interact with these mountpoints.

Reviewer 3: ```suggestion
An administrator logged in to the host OS wishing to obtain information about the Kubernetes system for debugging or auditing purposes will need to be aware of the new mount namespace. A shell originating in a container, such as `oc debug` will already be inside the Kubernetes namespace. A shell originating in an SSH session will be in the default namespace, and an administrator will need to run the `kubensenter` script to view or interact with these mountpoints:
```

Author: Done.

Issue the following command:

> [source,yaml]

Reviewer 3: Not yaml, but bash?

Reviewer 3: ```suggestion
[source,terminal]
```

----

> $ nsenter --mount=/run/container-mount-namespace/mnt ... <parameters ...> ...

Reviewer 3: This isn't right; it refers to the old implementation.  Will update on Monday.

Reviewer 3: ```suggestion
$ kubensenter <parameters ...> ...
```

Author: Updated.

> $ nsenter nsenter --mount=/run/container-mount-namespace/mnt ... <parameters ...> ...

Reviewer 3: ```suggestion
$ nsenter --mount=/run/container-mount-namespace/mnt ... <parameters ...> ...
```

----

> Example output:

Reviewer 3: ```suggestion
.Example output
```

Author: Done.

> [source,yaml]

Reviewer 3: Not yaml but ... command output?

----

> output from $ nsenter --mount=/run/container-mount-namespace/mnt ... <parameters ...> here

Reviewer 3: Fix this with real output, once the command above is fixed

Reviewer 3: ```suggestion
kubensenter: Autodetect: kubens.service namespace found at /run/kubens/mnt
/dev/sda4 on / type xfs (rw,relatime,seclabel,attr2,inode64,logbufs=8,logbsize=32k,prjquota)                                                                                                                                                                                                                                  
/dev/sda4 on /etc type xfs (rw,relatime,seclabel,attr2,inode64,logbufs=8,logbsize=32k,prjquota)                                                                                                                                                                                                                               
/dev/sda4 on /usr type xfs (ro,relatime,seclabel,attr2,inode64,logbufs=8,logbsize=32k,prjquota)                                                                                                                                                                                                                               
/dev/sda4 on /sysroot type xfs (ro,relatime,seclabel,attr2,inode64,logbufs=8,logbsize=32k,prjquota)                                                                                                                                                                                                                           
/dev/sda4 on /sysroot/ostree/deploy/rhcos/var type xfs (rw,relatime,seclabel,attr2,inode64,logbufs=8,logbsize=32k,prjquota)                                                                                                                                                                                                   
sysfs on /sys type sysfs (rw,nosuid,nodev,noexec,relatime,seclabel)                                                                                                                                                                                                                                                           
securityfs on /sys/kernel/security type securityfs (rw,nosuid,nodev,noexec,relatime)                                                                                                                                                                                                                                          
tmpfs on /sys/fs/cgroup type tmpfs (ro,nosuid,nodev,noexec,seclabel,mode=755)                                                                                                                                                                                                                                                 
cgroup on /sys/fs/cgroup/systemd type cgroup (rw,nosuid,nodev,noexec,relatime,seclabel,xattr,release_agent=/usr/lib/systemd/systemd-cgroups-agent,name=systemd)
...
```

Author: Done.

> ----

Reviewer 3: ```suggestion
----

To start a new interactive shell inside the Kubernetes namespace, run the `kubensenter` script without any arguments:

[source,terminal]
----
$ sudo kubensenter
----

.Example output
[source,terminal]
----
kubensenter: Autodetect: kubens.service namespace found at /run/kubens/mnt
#
----
```

> output from $ nsenter nsenter --mount=/run/container-mount-namespace/mnt ... <parameters ...> here

Reviewer 3: ```suggestion
output from $ nsenter --mount=/run/container-mount-namespace/mnt ... <parameters ...> here
```

Author: Done.

// Module included in the following assemblies:
//
// * scalability_and_performance/optimizing-cpu-usage.adoc

[id="supporting-encapsulation_{context}"]
= Inspecting encapsulated namespaces

> As an OpenShift support engineer or developer, you want to inspect Kubernetes-specific mountpoints.

Author: Went with "might want".

> An administrator logged in to the host OS wishing to obtain information about the Kubernetes system for debugging or auditing purposes will need to be aware of the new mount namespace. A shell originating in a container, such as `oc debug` will already be inside the Kubernetes namespace. A shell originating in an SSH session will be in the default namespace, and an administrator will need to run the `kubensenter` script as root to view or interact with these mountpoints.

Author: Done.

> The `kubensenter` script is aware of the state of the mount encapsulation feature status, and is safe to run even if encapsulation is not enabled. In that case it will simply execute the shell or requested command in the default mount namespace.

Author: Done.

> 

Author: Done.

. Open a remote shell to the cluster host or `oc debug node/<node-name>`.

. To run a single command inside the Kubernetes namespace, provide the command and any arguments to the `kubensenter` script.  For example, to run the `findmnt` command inside the Kubernetes namespace, issue the following command:
[source,terminal]
----

> $ sudo kubensenter findmnt

Reviewer 2: If we are running inside the cluster using `oc debug node`, the bash prompt will look like this:

```suggestion
sh-4.4# sudo kubensenter findmnt
```

Perhaps you get the dollar prompt when logged in via SSH? @lack?

Author: $ was an SME edit, leaving as is for now.

----

.Example output
[source,terminal]
----
  kubensenter: Autodetect: kubens.service namespace found at /run/kubens/mnt
  TARGET                                                                                                                                                   SOURCE                 FSTYPE     OPTIONS
  /                                                                                                                                                        /dev/sda4[/ostree/deploy/rhcos/deploy/32074f0e8e5ec453e56f5a8a7bc9347eaa4172349ceab9c22b709d9d71a3f4b0.0]
  |                                                                                                                                                                               xfs        rw,relatime,seclabel,attr2,inode64,logbufs=8,logbsize=32k,prjquota
                             shm                    tmpfs
  ...

> ----

Author: Done.

To start a new interactive shell inside the Kubernetes namespace, run the `kubensenter` script without any arguments:
[source,terminal]
----

> $ sudo kubensenter

Reviewer 2: If we are running inside the cluster using `oc debug node`, the bash prompt will look like this:

```suggestion
sh-4.4# sudo kubensenter
```

----

.Example output
[source,terminal]
----
kubensenter: Autodetect: kubens.service namespace found at /run/kubens/mnt

> ----

Author: Done.

// Module included in the following assemblies:
//
// * scalability_and_performance/optimizing-cpu-usage.adoc

[id="supporting-encapsulation_{context}"]
= Supporting encapsulation

As an OpenShift support engineer or developer, you want to inspect Kubernetes-specific mountpoints.

> == Inspecting the encapsulated namespaces

Reviewer 2: Since this is a module, we don't use second level headings, see: https://github.com/openshift/openshift-docs/blob/main/contributing_to_docs/doc_guidelines.adoc#writing-procedures. Perhaps the section could just be titled, for example, "Inspecting encapsulated namespaces"?

Author: Done. We need them as highlighted headings.

----
+
.Example output
[source,terminal]
----
  kubensenter: Autodetect: kubens.service namespace found at /run/kubens/mnt
  TARGET                                                                                                                                                   SOURCE                 FSTYPE     OPTIONS
  /                                                                                                                                                        /dev/sda4[/ostree/deploy/rhcos/deploy/32074f0e8e5ec453e56f5a8a7bc9347eaa4172349ceab9c22b709d9d71a3f4b0.0]
  |                                                                                                                                                                               xfs        rw,relatime,seclabel,attr2,inode64,logbufs=8,logbsize=32k,prjquota
                             shm                    tmpfs
  ...
----

To start a new interactive shell inside the Kubernetes namespace, run the `kubensenter` script without any arguments:
[source,terminal]
----
$ sudo kubensenter
----
+
.Example output
[source,terminal]
----
kubensenter: Autodetect: kubens.service namespace found at /run/kubens/mnt
----

> == Running additional services in the encapsulated namespace

Reviewer 2: This should be a separate procedure module based on https://github.com/openshift/openshift-docs/blob/main/contributing_to_docs/doc_guidelines.adoc#writing-procedures

Author: OK will split out and rename the previous as suggested.

Author: Done.

// Module included in the following assemblies:
//
// * scalability_and_performance/optimizing-cpu-usage.adoc

[id="supporting-encapsulation_{context}"]
= Supporting encapsulation

As an OpenShift support engineer or developer, you want to inspect Kubernetes-specific mountpoints.

> == Locating the encapsulated namespaces

Reviewer 3: I think we should rename this section "Running additional services in the encapsulated namespace" and move it AFTER the "Inspecting the encapsulated namespace" section below.

> Any monitoring tool that relies on the ability to run in the host OS and have visibility of mountpoints created by Kubelet, the Container Runtime, or containers themselves, will need to enter the container mount namespace in order to see these mountpoints. The provided `kubensenter` script will execute another command inside the Kubernetes mountpoint and can be used to adapt any existing tools.

Reviewer 2: Where is the provided `kubensenter` script? Note we cannot link to external git repos in the docs.

Reviewer 3: It's there already!  Every 4.12 (and later) OpenShift system will have the script pre-installed at `/usr/bin/kubensenter`, and this is true for both new deployments and upgrades to 4.12.

It's installed by the same RPM that installs the `kubelet` executable.

Author: As per Jim, it's installed with the product.

The `kubensenter` script is aware of the state of the mount encapsulation feature status, and is safe to run even if encapsulation is not enabled. In that case it will simply execute the provided command in the default mount namespace.
For example, if a systemd service needs to run inside the new Kubernetes mount namespace, you should edit the service file and prefix the `ExecStart=` command line with `kubensenter`.

[source,terminal]
----
[Unit]
Description=Example service
[Service]
ExecStart=/usr/bin/kubensenter /path/to/original/command arg1 arg2
----

== Inspecting the encapsulated namespaces

> An administrator logged in to the host OS wishing to obtain information about the Kubernetes system for debugging or auditing purposes will need to be aware of the new mount namespace. A shell originating in a container, such as `oc debug` will already be inside the Kubernetes namespace. A shell originating in an SSH session will be in the default namespace, and an administrator will need to run the `kubensenter` script as root to view or interact with these mountpoints.

Reviewer 2: So you only need this kubensenter script when SSHing into the cluster? not `oc debug`?

Reviewer 3: Correct.  Because the `oc debug` shell is in a container, it is already inside the Kubernetes mountpoint.

However, for ease in scripting it's fine to always run it - If you're already inside the namespace there's no harm in running it again, and if the feature is disabled there's also no harm in running it (in that it does nothing).

Author: Will discuss with Jim.

The `kubensenter` script is aware of the state of the mount encapsulation feature status, and is safe to run even if encapsulation is not enabled. In that case it will simply execute the shell or requested command in the default mount namespace.

To run a single command inside the Kubernetes namespace, provide the command and any arguments to the `kubensenter` script.  For example, to run the `findmnt` command inside the Kubernetes namespace, issue the following command:
[source,terminal]
----

> $ sudo kubensenter findmnt

Reviewer 2: Where are we running this from? inside the cluster host? I think it would be good to add steps before this command to clearly outline where you run this command from. ie, something like: 

1. Open a remote shell to the cluster host or `oc debug node/<node-name>`

...

Reviewer 3: In this case, it would be more important if the admin had SSHd in instead of `oc debug`, but I don't know if that's something we want to encourage users to do?  It is important for people to know, if they do SSH in, though.

Author: Will resolve with Jim.

  | |-/run/containers/storage/overlay-containers/35f65a7e081f135c3809350240e94cf6113b6e283f34191f8a5976041a52430e/userdata/shm                             shm                    tmpfs      rw,nosuid,nodev,noexec,relatime,context="system_u:object_r:container_file_t:s0:c284,c530",size=65536k
  | |-/run/utsns/f6f4d4cd-a701-462b-87dd-ec645b49f6b0                                                                                                      nsfs[uts:[4026532370]] nsfs       rw,seclabel
  | |-/run/ipcns/f6f4d4cd-a701-462b-87dd-ec645b49f6b0                                                                                                      nsfs[ipc:[4026532371]] nsfs       rw,seclabel
  | |-/run/netns/f6f4d4cd-a701-462b-87dd-ec645b49f6b0                                                                                                      nsfs[net:[4026531992]] nsfs       rw,seclabel
  | |-/run/containers/storage/overlay-containers/ec566c5b5e84eebe1f2e0f9259ac0096e87f6e49041f3f136379f7efbe4e74f2/userdata/shm                             shm                    tmpfs      rw,nosuid,nodev,noexec,relatime,context="system_u:object_r:container_file_t:s0:c294,c852",size=65536k

>   | |-/run/containers/storage/overlay-containers/ba6dc1dd63e5078fdb46b4a492b2172d63b10d26a4e8fb613ab301658ef3d1b2/userdata/shm                             shm                    tmpfs

Reviewer 3: This is not the full output.  The full output on an OpenShift cluster is very very long.  Should we add an ellipsis to indicate there's more they would see in the real world?
```suggestion
  | |-/run/containers/storage/overlay-containers/ba6dc1dd63e5078fdb46b4a492b2172d63b10d26a4e8fb613ab301658ef3d1b2/userdata/shm                             shm                    tmpfs
...
```

Author: Yes, let's go with the ellipse.

----

To start a new interactive shell inside the Kubernetes namespace, run the `kubensenter` script without any arguments:
[source,terminal]
----
$ sudo kubensenter
----
+
.Example output
[source,terminal]
----
kubensenter: Autodetect: kubens.service namespace found at /run/kubens/mnt

> #

Reviewer 3: ```suggestion
#
----
```
I think this is missing the closing `----`?

Reviewer 2: ^^ Tony this is your build failing issue

Author: It sure was. Fixed.

Author: Fixed.
